### PR TITLE
tiles: replace poh and bank and store

### PR DESCRIPTION
**Pull request by mmcgee-jump**

### CODE REVIEW

**Comment by mmcgee-jump** [comment]
Few things left,

- [x] unprivileged_init functions need to be written
- [x] the actual sanitizer code for the SanitizedTransaction creation needs to be ported over
- [x] Completely untested and likely a lot of issues
- [x] The grace ticks code maybe needs to be ported over and added to the poh leader publish timing?
- [x] The leader update frags from PoH need to be ignored by the shredder
- [x] More metrics for the bank tile, particularly around transaction execution, and why things failed to execute etc
- [x] Busy fseq logic needs to change, so counter is consistent across tiles

File: src/app/fdctl/run/tiles/fd_poh.c [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+  /* Indicate to pack tile we are done processing the transactions so it
+     can pack new microblocks using these accounts.  DO NOT USE THE
+     SANITIZED TRANSACTIONS AFTER THIS POINT, THEY ARE NOT LONGER VALID. */
+  FD_COMPILER_MFENCE();
+  *(ctx->pack_busy[ target_bank_idx ]) = 0UL;
+
+  /* Publish the microblock. */
+  fd_entry_batch_meta_t * meta = (fd_entry_batch_meta_t *)dst;
+  meta->parent_offset = (ctx->hash_cnt/HASHES_PER_SLOT) - (ctx->reset_slot_hash_start/HASHES_PER_SLOT);
+  meta->bank_max_tick_height = (ctx->next_leader_slot_hash_start/HASHES_PER_TICK)+1UL;
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/tiles/fd_poh.c:542]: Might not be right? What is tick height here.

File: src/disco/bank/fd_bank_abi.c [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+};
+
+static int
+is_builtin_key_or_sysvar( uchar * key ) {
+  for( ulong i=0; i<sizeof(BUILTIN_SYSVARS_AND_PROGRAMS)/sizeof(*BUILTIN_SYSVARS_AND_PROGRAMS); i++ ) {
+    if( !memcmp( key, BUILTIN_SYSVARS_AND_PROGRAMS[i], 32UL ) ) return 1;
+  }
+
+  return 0;
+}
```
Comment (ptaffet-jump) [inline src/disco/bank/fd_bank_abi.c:235]: You are right that we still need this here to check the accounts in the address lookup table. The list in `fd_pack.c` is almost correct. I'll work on fixing it

**Review by ptaffet-jump — COMMENTED** [review]
In progress

File: src/disco/bank/fd_bank_abi.c [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+
+  out_txn->is_simple_vote_tx = !!is_simple_vote;
+
+  if( FD_LIKELY( txn->transaction_version==FD_TXN_VLEGACY ) ) {
+    out_txn->message.tag = 0UL;
+    out_txn->message.legacy.is_writable_account_cache_cnt = txn->acct_addr_cnt;
+    out_txn->message.legacy.is_writable_account_cache_cap = txn->acct_addr_cnt;
+    out_txn->message.legacy.is_writable_account_cache = out_sidecar;
+    int _is_upgradeable_loader_present = is_upgradeable_loader_present_legacy( txn, payload );
+    for( ushort i=0; i<txn->acct_addr_cnt; i++ ) {
```
[the remark this answers is not in the document]
Comment (mmcgee-jump) [reply src/disco/bank/fd_bank_abi.c:295]: I don't think it's much clearer here, and anyway we need to iterate the rest of the accounts to set the value to 0.

**Review by ptaffet-jump — COMMENTED** [review]
Still in progress

File: src/app/fdctl/run/run1.c [hunk 4]
```diff
@@ -50,20 +50,25 @@ tile_main( void * _args ) {
...
-  fd_log_private_group_id_set( pid );
-  fd_log_private_thread_id_set( pid );
-  fd_log_private_stack_discover( FD_TILE_PRIVATE_STACK_SZ,
-                                 &fd_tile_private_stack0, &fd_tile_private_stack1 );
+  fd_log_cpu_set( NULL );
+  fd_log_private_tid_set( pid );
+  char thread_name[ FD_LOG_NAME_MAX ] = {0};
+  snprintf1( thread_name, FD_LOG_NAME_MAX-1UL, "%s:%lu", fd_topo_tile_kind_str( tile->kind ), tile->kind_id );
+  fd_log_thread_set( thread_name );
+  // fd_log_private_stack_discover( FD_TILE_PRIVATE_STACK_SZ,
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/run1.c:58]: Need to restore this

File: src/disco/bank/fd_bank_abi.c [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    for( ushort i=0; i<txn->addr_table_adtl_writable_cnt; i++ ) {
+      int is_writable = !is_builtin_key_or_sysvar( (uchar*)(out_txn->message.v0.loaded_addresses.owned.writable + i) ) &&
+                        (!is_key_called_as_program( txn, (ushort)(txn->acct_addr_cnt+i) ) || _is_upgradeable_loader_present);
+      out_txn->message.v0.is_writable_account_cache[ txn->acct_addr_cnt+i ] = !!is_writable;
+    }
+    for( ushort i=0; i<txn->addr_table_adtl_cnt-txn->addr_table_adtl_writable_cnt; i++ ) {
+      out_txn->message.v0.is_writable_account_cache[ txn->acct_addr_cnt+txn->addr_table_adtl_writable_cnt+i ] = 0;
+    }
+
+    out_txn->message.v0.message.owned.account_keys_cnt = txn->acct_addr_cnt;
```
Comment (ptaffet-jump) [inline src/disco/bank/fd_bank_abi.c:364]: This code would get a lot cleaner if you declared a pointer to `owned` as a local variable. That would require giving the type a name though.

**Review by ptaffet-jump — COMMENTED** [review]
Still in progress, but fully reviewed the ABI part

File: src/app/fdctl/run/tiles/fd_poh.c [hunk 6]
```diff
@@ -0,0 +1,1192 @@
...
+fd_ext_poh_write_unlock( void ) {
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( fd_poh_returned_lock ) = 0UL;
+}
+
+/* The PoH tile needs to interact with the Solana Labs address space to
+   do certain operations that Firedancer hasn't reimplemented yet, a.k.a
+   transaction execution.  We have Solana Labs export some wrapper
+   functions that we call into during regular tile execution.  These do
+   not need any locking, since they are called serially from the single
```
[the remark this answers is not in the document]
Comment (mmcgee-jump) [reply src/app/fdctl/run/tiles/fd_poh.c:448]: We will find out. Technically in Solana Labs, commit_txns happens from the banking stage so they are not serial.  If it ends up being too slow, we can push this to a new `commit` tile that sits behind POH tile and have n of them.

File: src/disco/bank/fd_bank_abi.c [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+  ulong signatures_cap;
+  ulong signatures_cnt;
+
+  sanitized_txn_abi_message_t message;
+
+  uchar message_hash[ 32 ];
+  uchar is_simple_vote_tx;
+};
+
+FD_STATIC_ASSERT( sizeof(struct fd_bank_abi_txn_private) == 248UL, "messed up size" );
```
[the remark this answers is not in the document]
Comment (mmcgee-jump) [reply src/disco/bank/fd_bank_abi.c:188]: I just added a assert_sizeof check to the Rust side which should hopefully be sufficient.

**Review by ptaffet-jump — COMMENTED** [review]
Pending from yesterday. Sorry, just noticed.

File: src/app/fdctl/config.c [hunk 8]
```diff
@@ -517,19 +519,20 @@ topo_initialize( config_t * config ) {
...
   LINK( 1,                                FD_TOPO_LINK_KIND_STAKE_TO_OUT,    FD_TOPO_WKSP_KIND_STAKE_OUT,    128UL,                                    32UL + 40200UL * 40UL,  1UL );
   LINK( 1,                                FD_TOPO_LINK_KIND_PACK_TO_BANK,    FD_TOPO_WKSP_KIND_PACK_BANK,    128UL,                                    USHORT_MAX,             1UL );
-  LINK( 1,                                FD_TOPO_LINK_KIND_POH_TO_SHRED,    FD_TOPO_WKSP_KIND_BANK_SHRED,   128UL,                                    FD_POH_SHRED_MTU,       1UL );
-  LINK( 1,                                FD_TOPO_LINK_KIND_CRDS_TO_SHRED,   FD_TOPO_WKSP_KIND_BANK_SHRED,   128UL,                                    8UL  + 40200UL * 38UL,  1UL );
-  /* See long comment in fd_shred_tile.c for an explanation about the size of this dcache. */
+  LINK( config->layout.bank_tile_count,   FD_TOPO_LINK_KIND_BANK_TO_POH,     FD_TOPO_WKSP_KIND_BANK_POH,     128UL,                                    USHORT_MAX,             1UL );
+  LINK( 1,                                FD_TOPO_LINK_KIND_POH_TO_SHRED,    FD_TOPO_WKSP_KIND_POH_SHRED,    128UL,                                    USHORT_MAX,             1UL );
```
[the remark this answers is not in the document]
Comment (ptaffet-jump) [reply src/app/fdctl/config.c:533]: These could be tightened, but it's fine

**Review by lidatong — APPROVED** [review]

**Pull request merged**